pkg/autoscaler/stats_scraper.go

```diff
@@ -52,48 +60,119 @@ var cacheDisabledClient = &http.Client{
+	count int
+}
+
 // ServiceScraper scrapes Revision metrics via a K8S service by sampling. Which
 // pod to be picked up to serve the request is decided by K8S. Please see
 // https://kubernetes.io/docs/concepts/services-networking/network-policies/
 // for details.
 type ServiceScraper struct {
-	httpClient *http.Client
-	url        string
-	metricKey  string
-	logger     *zap.SugaredLogger
+	httpClient      *http.Client
+	endpointsLister corev1listers.EndpointsLister
+	url             string
+	namespace       string
+	revisionService string
+	metricKey       string
+	cachedReadyPods readyPods
 }
 
 // NewServiceScraper creates a new StatsScraper for the Revision which
 // the given Metric is responsible for.
-func NewServiceScraper(metric *Metric, logger *zap.SugaredLogger) (*ServiceScraper, error) {
-	return newServiceScraperWithClient(metric, logger, cacheDisabledClient)
+func NewServiceScraper(metric *Metric, dynamicConfig *DynamicConfig, endpointsInformer corev1informers.EndpointsInformer) (*ServiceScraper, error) {
+	return newServiceScraperWithClient(metric, dynamicConfig, endpointsInformer, cacheDisabledClient)
 }
 
-func newServiceScraperWithClient(metric *Metric, logger *zap.SugaredLogger, httpClient *http.Client) (*ServiceScraper, error) {
+func newServiceScraperWithClient(
+	metric *Metric,
+	dynamicConfig *DynamicConfig,
+	endpointsInformer corev1informers.EndpointsInformer,
+	httpClient *http.Client) (*ServiceScraper, error) {
+	if metric == nil {
+		return nil, errors.New("empty point of Metric")
+	}
+	if dynamicConfig == nil {
+		return nil, errors.New("empty point of DynamicConfig")
+	}
+	if endpointsInformer == nil {
+		return nil, errors.New("empty interface of EndpointsInformer")
+	}
+	if httpClient == nil {
+		return nil, errors.New("empty point of HTTP client")
+	}
 	revName := metric.Labels[serving.RevisionLabelKey]
 	if revName == "" {
 		return nil, fmt.Errorf("no Revision label found for Metric %s", metric.Name)
 	}
 
 	serviceName := reconciler.GetServingK8SServiceNameForObj(revName)
 	return &ServiceScraper{
-		httpClient: httpClient,
-		url:        fmt.Sprintf("http://%s.%s:%d/metrics", serviceName, metric.Namespace, v1alpha1.RequestQueueMetricsPort),
-		metricKey:  NewMetricKey(metric.Namespace, metric.Name),
-		logger:     logger,
+		httpClient:      httpClient,
+		endpointsLister: endpointsInformer.Lister(),
+		url:             fmt.Sprintf("http://%s.%s:%d/metrics", serviceName, metric.Namespace, v1alpha1.RequestQueueMetricsPort),
+		metricKey:       NewMetricKey(metric.Namespace, metric.Name),
+		namespace:       metric.Namespace,
+		revisionService: reconciler.GetServingK8SServiceNameForObj(revName),
+		cachedReadyPods: readyPods{},
 	}, nil
 }
 
 // Scrape call the destination service then send it
 // to the given stats chanel
-func (s *ServiceScraper) Scrape(statsCh chan<- *StatMessage) {
+func (s *ServiceScraper) Scrape(ctx context.Context, statsCh chan<- *StatMessage) {
+	logger := logging.FromContext(ctx)
```

Comment:
I wonder if we should make `Scrape` return an `error` and log in the method calling it to prevent creating loggers (and passing context just to create loggers)

Reply:
I made this function as the same format of `Scale` function. If we want to send no pods log for debugging, we can't return it as error.

Reply:
We can have logger as a `ServiceScraper` member, fwiw.